File: 08.Boyer-MoreeAlgo.py
# The Boyer-Moore Majority Vote Algorithm is a linear-time algorithm used 
# to find the majority element in an array (the element that appears more than ⌊n/2⌋ times).
# It works by maintaining a candidate and a counter while traversing the array in one pass.

# Explanation

# The idea is to keep track of a candidate element that could be the majority.

# We maintain a count:

# If count becomes 0, we pick the current element as the new candidate.

# If the current element equals the candidate → increment count.

# Otherwise → decrement count.

# At the end, the candidate is the potential majority element (needs a second pass to verify in some variations).

def majority_element(nums):
    candidate, count = None, 0

    for num in nums:
        if count == 0:
            candidate = num
        
        if num == candidate:
            count += 1
        else:
            count -= 1

    return candidate

arr = [2, 2, 1, 1, 1, 2, 2]
result = majority_element(arr)
print("Majority Element:", result)

        

File: 08.Boyer-MoreeAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio        */
/* -----------------------------------------------------------------------  */

import java.io.*;

public class BoyerMoore {
    
    // Function to find the majority element
    public static int majorityElement(int[] nums) {
        // Step 1: Initialize candidate and count
        int candidate = -1; 
        int count = 0;

        // Step 2: Traverse through the array
        for (int num : nums) {
            if (count == 0) {
                // If count is 0, choose current element as candidate
                candidate = num;
            }
            if (num == candidate) {
                // If element matches candidate, increase count
                count++;
            } else {
                // Otherwise, decrease count
                count--;
            }
        }

        // Step 3: Return candidate (majority element)
        return candidate;
    }

    // Main function to test the algorithm
    public static void main(String[] args) throws IOException {
        // Example input array
        int[] arr = {2, 2, 1, 1, 1, 2, 2};

        // Call the function and store result
        int result = majorityElement(arr);

        // Print the majority element
        System.out.println("Majority Element: " + result);
    }
}
